### ISSUE TITLE
Refund Functions with Stripe Express - Full Bugs Report

### ISSUE THREAD

**Issue by Sophie-2e** [getdokan/dokan#2009]
### Bug Description

Previously, below were the issues I had before updating Dokan, with the refund function with Stripe Express:

1. Partial Refunds with Stripe: this option "the partial refund" option has been deleted and included in the refund function directly, so it was working in Stripe, also in admin.
a. But in the seller dashboard, earnings from the order and refunded amount were not updated so for sellers, it seems that it has not worked while it did.
b. Also, vendors did not receive any notification that the refund had been made, nor did customers/guest customers.

2. Full Refunds with Stripe: it was working, and refunds got directly accepted without admin needing to accept or not the refund.
a. But in the seller dashboard, earnings from the order amount were not updated, it should even become negative as the commission is not refunded.

3. Manual Refund: Refund should not be manual, the seller has the option to do a manual refund but they don't have customer details as we hide it from them. We don't want this option, it should be hidden for all sellers if the option to hide the customers' details is activated in Dokan Admin Settings.

=========

Here are the issues with the latest Dokan Versions:
1. Partial Refunds with Stripe: if we use the "Qty" field to initiate a refund, it does not work at all, even if you select only one article, it automatically calculates a refund amount which is higher than the order amount itself so which will be automatically cancelled by Stripe: https://ibb.co/f2dgRNj

2. When customers request a refund: an email is sent to vendor (Dokan Send Refund Request to Vendor) but the url link is not working, it is not the right url: https://ibb.co/nzFTx8s
https://ibb.co/P6DQXHt
The link should be: seller-dashboard/orders/?order_id=25199&_wpnonce=7ddeb07219
But it is: seller-dashboard/orders/?order_id=25199&_wpnonce=4db4809304

3. In the customer dashboard / My Account, it is still shown as "RMA" while "Return Request" is a clearer term, even for customers. The best would be "Returns & Refunds": https://ibb.co/V2ckjxQ

4. The Return Request Tab in the Vendor Dashboard should also have the notification on it like the support tab has with the number of new ticket (): https://ibb.co/pwFTBDK

5. Request Coupon Credit: I don't understand when the email "Dokan Send Coupon To Customer" is sent to customer as for now this is the process:
- Customer got is order completed by vendor and can request a warranty via the order tab in My Account
- He request a store credit and it open a request in their RMA requests tab and in the Vendor Returns Request tab
- From there the only thing the vendor can do is send a message to the customer or change the status of the request while the customer can only send a message to the vendor. So there is no way to add the "Store Credit Coupon" and initiate the sending of the "Dokan Send Coupon To Customer". On top of it, 0 notification emails are sent to the customer, seller or admin when someone answers the request by message, so hard to keep up.
- it seems that the vendor needs to go to the Coupons tab > and create a new coupon but we cannot say that it is a coupon dedicated to a refund request so again the email "Dokan Send Coupon To Customer" is not sent. 
- in the coupon settings, there is a field about the Email Restrictions" but vendors do not have acess to the customers' email anyways...
So I really don't see when this email is sent. And the customer does not get any email when the vendor answers in the Request so they do not know what is going on, especially if they do not have an account on the website...

6. Manual Refund: Refund should not be manual, the seller has the option to do a manual refund but they don't have customer details as we hide it from them. We don't want this option, it should be hidden for all sellers if the option to hide the customers' details is activated in Dokan Admin Settings.

### Step To Reproduce

Test the refund finctionalities
functionalities

### Expected Behaviour

Cf Bug Description - already reported all that does not make sense and work in details so I let you figure out the best solution for this functionality.

### Actual Behaviour

Cf Bug Description 

### Additional Information

Cf Bug Description 

### Environment (please complete the following information)

```shell
1. Dokan Version: latest
2. Dokan Pro Version: latest
3. WooCommerce Version: latest
4. Browser all
5. Current Theme: Blocksy/Elementor does not matter, same results anyways
```

**Comment by imtiaz-pranto** [comment]
We have done an initial analysis of the issues, and here is what we have found:
1. This is actually how woocommerce handles the refund, as the total order is considered for refunds. Also from Stripe, there is no way to control which amount gets refunded from a combined order with automated refund. We suggest using suborders for partial refunds, which can minimize the issue. We will work on it to find a better solution.
2. Same as 1
3. We will consider adding a vendor capability setting in the admin panel to solve this issue. 
Issues
1. Valid Issue. Amount validation will be added for refunds to prevent this.
2. Valid Issue. Email URL will be checked.
3. Good Suggestion, We will consider changing the name for better understanding.
4. Good Suggestion. We can add a badge like other modules for better accessiblity. 
5. This is a complicated issue and we are working on it to find a good solution. 
6. Same as no 3 of refund issue.

**Comment by Sophie-2e** [comment]
Hello @imtiaz-pranto 

Difficult to understand what are your answers to the issues reported as we did not number them in the same way and you are not explaining what is the topic of your points. 

Anyways, just tested again since Dokan updated to 3.9 version. Here are the results of all issues which need to be fixed:


1. Partial Refunds with Stripe: if we use the "Qty" field to initiate a refund, it does not work at all, even if you select only one article, it automatically calculates a refund amount which is higher than the order amount itself so which will be automatically cancelled by Stripe: https://ibb.co/f2dgRNj It should be fixed and vendors should be able to use it, and even if it comes from Woocommerce, Dokan should then to block this option in the vendor order detail then because every vendor will use the quantity feature as it is a lot easier than calculated the VAT or product price of products individually.

2. When customers request a refund: an email is sent to vendor (Dokan Send Refund Request to Vendor) but the url link is still not working, it is not the right url: https://ibb.co/nzFTx8s
https://ibb.co/P6DQXHt
The link should be: seller-dashboard/orders/?order_id=25199&_wpnonce=7ddeb07219
But it is: seller-dashboard/orders/?order_id=25199&_wpnonce=4db4809304

3. In the customer dashboard / My Account, it is still shown as "RMA" while "Return Request" is a clearer term, even for customers. The best would be "Returns & Refunds": https://ibb.co/V2ckjxQ

4. The Return Request Tab in the Vendor Dashboard should also have the notification on it like the support tab has with the number of new tickets: https://ibb.co/pwFTBDK

5. Request Coupon Credit: The email "Dokan Send Coupon To Customer" is never sent to customers as for now this is no option for the vendor to send a store coupon credit via a customer support or refund request.

6. Manual Refund: Refund should not be manual, the seller has the option to do a manual refund but they don't have customer details as we hide it from them. We don't want this option, it should be hidden for all sellers if the option to hide the customers' details is activated in Dokan Admin Settings.

7. The customer has no idea of what is going on as they receive 0 email notifications about their refund request.

**Comment by Sophie-2e** [comment]
Hi @imtiaz-pranto, @nurul-umbhiya and @shams-sadek1981 

I see that you fixed the manual refund issue but all the other issues of the refund function are still there. Note that right now we cannot do partial or full refunds so we cannot do any refund. Here is the list of issues still there in Dokan 3.91 and Pro 3.9.2 and the new one about the full refund which is not possible anymore:

1. FULL AND PARTIAL REFUND is not working at all: 
a. with quantity field: still not working, keep having an error: https://ibb.co/FxxrtRD
b. with total + vat field: Keep having an error saying the refund amount is bigger than the order amount https://ibb.co/KLn9Lzt while it is not: https://ibb.co/Bn8bRyF

So now we cannot do a partial refund but a full refund either....so we cannot refund at the moment.

2. When customers request a refund: an email is sent to the vendor (Dokan Send Refund Request to Vendor) but the URL link is still not working, it is not the right URL: https://ibb.co/nzFTx8s
https://ibb.co/P6DQXHt
The link should be: seller-dashboard/orders/?order_id=25199&_wpnonce=7ddeb07219
But it is: seller-dashboard/orders/?order_id=25199&_wpnonce=4db4809304

3. In the customer dashboard / My Account, it is still shown as "RMA" while "Return Request" is a clearer term, even for customers. The best would be "Returns & Refunds": https://ibb.co/V2ckjxQ

4. The Return Request Tab in the Vendor Dashboard should also have the notification on it like the support tab has with the number of new tickets: https://ibb.co/pwFTBDK

5. The customer has no idea of what is going on as they receive 0 email notifications about their refund request when vendor reply to the RMA ticket.

6. Same if the customer replies to the refund ticket, the vendor has no idea as he does not received an email and there is no notification on the Return Request Tab in the Vendor Dashboard.

Can I have an update fast as this is really an emergency. Vendors need to be able to process refunds. Waiting for your news.

**Comment by nurul-umbhiya** [comment]
@shohag121 bhai can you please start working on this issue?

**Comment by shohag121** [comment]
@Mohaiminulislam1989 Bhai, Can you please provide the Email contents for Tasks 5 and 6?

**Comment by Sophie-2e** [comment]
@nurul-umbhiya, @shohag121  I forgot this: 

7. Request Coupon Credit: The email "Dokan Send Coupon To Customer" is never sent to customers as for now there is no option for the vendor to send a store coupon credit via a customer support or refund request so how does this work then?

Recap issue:
Request Coupon Credit: I don't understand when the email "Dokan Send Coupon To Customer" is sent to the customer, as for now this is the process:

- Customer got his order completed by the vendor and can request a warranty via the order tab in My Account
- He requests a store credit and it opens a request in their RMA requests tab and in the Vendor Returns Request tab
- From there the only thing the vendor can do is send a message to the customer or change the status of the request while the customer can only send a message to the vendor. So there is no way to add the "Store Credit Coupon" and initiate the sending of the "Dokan Send Coupon To Customer". On top of it, 0 notification emails are sent to the customer, seller or admin when someone answers the request by message, so hard to keep up.
- it seems that the vendor needs to go to the Coupons tab > and create a new coupon but we cannot say that it is a coupon dedicated to a refund request so again the email "Dokan Send Coupon To Customer" is not sent.
- in the coupon settings, there is a field about the Email Restrictions" but vendors do not have access to the customers' email anyway...

So I really don't see when this email is sent. And the customer does not get any email when the vendor answers the Request so they do not know what is going on, especially if they do not have an account on the website...

**Comment by shohag121** [comment]
@shashwatahalder01 Can you please check the issue number **7** ASAP and generate a detailed bug report?

cc: @shams-sadek1981 @Mohaiminulislam1989

**Comment by shashwatahalder01** [comment]
@shohag121 , @Sophie-2e  The request for coupon credit is functioning as intended, and it is working perfectly. You can watch a detailed video explanation at this [link](https://www.awesomescreenshot.com/video/22148381?key=ef075d5e3ff5728d74c3a9b1626cd8ec)

@Sophie-2e , "From there the only thing the vendor.........", At this point, the vendor has to change the status to "processing" in order to access the refund or send coupon option. It's worth noting that the vendor doesn't need to create a coupon manually for the coupon credit.

**Comment by Sophie-2e** [comment]
@shashwatahalder01 @shohag121 : oooh I got it, I was not changing the status of the RMA request so I was not seeing the coupon button 😅. So I tested everything, and email is sent, coupon works well. The only issue is that we are not able to fully refund the product, in your example you are refunding 30$ less than the full amount, and like this it works, but if you try fully refund the product, it does not work. I am thinking that maybe it is because the admin commission has been taken so maybe that is why the vendor does not have the full amount and cannot refund it fully? 🤔 Here is a video showing the issue :  https://share.vidyard.com/watch/Z926iJawtxrzZHBBcetKXq?

Let me know :)

**Comment by Sophie-2e** [comment]
@nurul-umbhiya @shohag121 this issue is back...Manual Refund > See option again even if Hide Customer Detail is enabled in Dokan Admin Settings. Just to let you know. Another fix that only last for a few month (was fixed in october 2023 in 3.9.1)
